potoken.py

```diff
@@ -0,0 +1,255 @@
+            visitor_data = post_data_json['context']['client']['visitorData']
+            potoken = post_data_json['serviceIntegrityDimensions']['poToken']
+        except (json.JSONDecodeError, TypeError, KeyError) as e:
+            logger.warning(f'failed to extract token from request: {type(e)}, {e}')
+            return None
+        token_info = {
+            'updated': int(time.time()),
+            'potoken': potoken,
+            'visitor_data': visitor_data
+        }
+        return token_info
+
+    async def _update(self) -> None:
+        try:
+            await asyncio.wait_for(self._perform_update(), timeout=600)
+        except asyncio.TimeoutError:
+            logger.error(f'update failed: hard limit timeout exceeded. Browser might be failing to start properly')
+
+    async def _perform_update(self) -> None:
+        if self._ongoing_update.locked():
+            logger.debug('update is already in progress')
+            return
+
+        async with self._ongoing_update:
+            logger.info(f'update started')
+            self._extraction_done.clear()
+
+            browser = await nodriver.start(headless=False, user_data_dir=self.profile_path)
+            tab = browser.main_tab
+            tab.add_handler(nodriver.cdp.network.RequestWillBeSent, self._send_handler)
+            await tab.get('https://www.youtube.com/embed/jNQXAC9IVRw')
+            player_clicked = await self._click_on_player(tab)
+            if player_clicked:
+                await self._wait_for_handler()
+            await tab.close()
+            browser.stop()
+
+    @staticmethod
+    async def _click_on_player(tab) -> bool:
+        try:
+            player = await tab.select("#movie_player", 10)
+        except asyncio.TimeoutError:
+            logger.warning(f'update failed: unable to locate video player on the page')
+            return False
+        else:
+            await player.click()
+            return True
+
+    async def _wait_for_handler(self) -> bool:
+        try:
+            await asyncio.wait_for(self._extraction_done.wait(), timeout=30)
+        except asyncio.TimeoutError:
+            logger.warning(f'update failed: timeout waiting for outgoing API request')
+            return False
+        else:
+            logger.info('update was succeessful')
+            return True
+
+    async def _send_handler(self, event: nodriver.cdp.network.RequestWillBeSent):
+        if not event.request.method == 'POST':
+            return
+        if not '/youtubei/v1/player' in event.request.url:
+            return
+        token_info = self._extract_token(event.request)
+        if token_info is None:
+            return
+        logger.info(f'new token: {token_info}')
+        self._token_info = token_info
+        self._extraction_done.set()
+
+
+class ThreadingWSGIServer(WSGIServer, ThreadingMixIn):
+    """Thread per request HTTP server."""
+    daemon_threads = True
+
+
+class PotokenServer:
```

Comment:
Server logic should probably be separated from  extractor logic. Possible structure
```
- session_generator/
   - __init__ - Contains CLI entry point & server entry point
   - server
   - extractor
```

Reply:
I would like to postpone packaging this as a module until it comes to publishing it to pypi.
It's quite handy to have it as a single file that can be run from any directory without installing, so I would rather keep it in this state while possible.

Reply:
https://github.com/iv-org/youtube-trusted-session-generator/pull/6#discussion_r1765773287